### PR TITLE
v1 and v2 token Audience claim verification support

### PR DESCRIPTION
**Pull request by idg-sam**
- v1 and v2 token audience claim verification support
- README revisions

### CODE REVIEW

File: msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+        this.resourceId = resourceId;
+        // prepend https:// and append tenantId. And then duplicate this and append /v2.0
+        this.issuers = this.addHttpsTenantIdAndV2Endpoints(issuers, issuerTenant);
+    }
+
+    private String[] addHttpsTenantIdAndV2Endpoints(final String[] issuers, final String issuerTenant) {
+        final String[] issuersV2Included = new String[issuers.length * 2];
+        for (int i = 0; i < issuers.length; i++) {
+            issuers[i] = "https://" + issuers[i] + "/" + issuerTenant + "/";
```
[the remark this answers is not in the document]
Comment (SomkaPe) [reply msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java:34]: I would explicitly define format of issue using constant like : 
V2_ISSUER_FORMAT="https://%s/%s/v2"
+
String.format(V2_ISSUER_FORMAT, issuer, tenant)

which might be cleaner and easy to read

File: msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.security.oauth2.provider.token.store.JwtClaimsSetVerifier;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import java.util.Arrays;
+import java.util.Map;
+
+public class AADClaimsVerifier implements JwtClaimsSetVerifier {
+    private static final String ISS_CLAIM = "iss";
+
+    private final String[] issuers;
```
Comment (SomkaPe) [inline msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java:12]: I would use set here, issuers should be unique

File: AppCreationScripts/sample.json [hunk 3]
```diff
@@ -1,54 +1,71 @@
 {
   "Sample": {
```
Comment (kalyankrishna1) [inline AppCreationScripts/sample.json:2]: Sam, work with Dogan to see why the Codegenerator wont work for this sample.json file

Comment (idg-sam) [reply AppCreationScripts/sample.json:2]: Sure thing. The generator doesn't work properly on this file.

I just committed a working app creation and cleanup script that I hacked together from the output of the generator though. It seems to at least make the app creation script files, but I had to modify them to remove some extra spaces.

File: msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (idg-sam) [reply msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java:34]: Thanks Petro, I'm changing it now.

File: msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java (hunk 2 above)
Comment (kalyankrishna1) [reply msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java:12]: As a class I would advise we follow the .net middleware team's Api surface ,where they allow both a single issuer and multiple issuers to be set. This is extremely useful for multi-tenant apps, who use the array to restrict the tokens presented to them from a select number of tenants. Note: My knowledge of Java is rudimentary, and I'm assuming this class will be used to verify tokens

File: msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (idg-sam) [reply msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java]: I've updated the implementation.

File: msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    private String[] addHttpsTenantIdAndV2Endpoints(final String[] issuers, final String issuerTenant) {
+        final String[] issuersV2Included = new String[issuers.length * 2];
+        for (int i = 0; i < issuers.length; i++) {
+            issuers[i] = "https://" + issuers[i] + "/" + issuerTenant + "/";
+            issuersV2Included[i] = issuers[i];
+            issuersV2Included[issuers.length + i] = issuers[i] + "v2.0";
+        }
+        return issuersV2Included;
```
[the remark this answers is not in the document]
Comment (idg-sam) [reply msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java]: The issuer claim format changes in v1 vs v2 tokens, as does the aud claim.

File: msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java (hunk 2 above)
Comment (idg-sam) [reply msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/AADClaimsVerifier.java]: Changed to set. Added possibility for multiple accepted issuing tenants.

File: AppCreationScripts/sample.json (hunk 3 above)
Comment (idg-sam) [reply AppCreationScripts/sample.json:2]: So the only thing it is not properly generating at this point is the apps.json file (which is only required for quickstart projects, according to Dogan). Dogan has also confirmed that the config and cleanup script included here are able to get the project running fully.

**Pull request merged**